src/bin/muxer/avformat.rs

```diff
@@ -0,0 +1,123 @@
+// the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+// was not distributed with this source code in the LICENSE file, you can
+// obtain it at www.aomedia.org/license/software. If the Alliance for Open
+// Media Patent License 1.0 was not distributed with this source code in the
+// PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+
+#[cfg(feature = "ffmpeg-sys")]
+use ffmpeg_sys::*;
+
+#[allow(unused_imports)]
+use super::Muxer;
+#[allow(unused_imports)]
+use std::ffi::CString;
+#[allow(unused_imports)]
+use std::io;
+#[allow(unused_imports)]
+use std::mem;
+#[allow(unused_imports)]
+use std::ptr;
+
+#[cfg(feature = "ffmpeg-sys")]
+pub struct AvformatMuxer {
+  context: *mut AVFormatContext,
+  duration: i64
+}
+
+#[cfg(feature = "ffmpeg-sys")]
+impl AvformatMuxer {
+  pub fn open(path: &str) -> Box<dyn Muxer> {
+    unsafe {
+      let mut context = ptr::null_mut();
+      let p = CString::new(path).expect("convert to cstring failed");
+      match avformat_alloc_output_context2(
+        &mut context,
+        ptr::null_mut(),
+        ptr::null(),
+        p.as_ptr() as *const i8
+      ) {
+        0 => {
+          let mut io = ptr::null_mut();
+          let ret =
+            avio_open(&mut io, p.as_ptr() as *const i8, AVIO_FLAG_WRITE);
+          if ret < 0 {
+            panic!("open ouput file {} failed", path);
+          }
+          (*context).pb = io;
+          Box::new(AvformatMuxer { context, duration: 0 })
+        }
+        e => panic!("open ouput failed, error = {}", e)
+      }
+    }
+  }
+}
+
+#[cfg(feature = "ffmpeg-sys")]
+impl Muxer for AvformatMuxer {
+  fn write_header(
+    &mut self, width: usize, height: usize, framerate_num: usize,
+    framerate_den: usize
+  ) {
+    unsafe {
+      let stream = avformat_new_stream(self.context, ptr::null_mut());
+      if stream.is_null() {
+        panic!("new stream failed");
+      }
+      let param = (*stream).codecpar;
+      (*param).codec_type = AVMediaType::AVMEDIA_TYPE_VIDEO;
+      (*param).codec_id = AVCodecID::AV_CODEC_ID_AV1;
+      (*param).width = width as i32;
+      (*param).height = height as i32;
+      let ret = avformat_write_header(self.context, ptr::null_mut());
+      if ret < 0 {
+        panic!("write header failed error = {}", ret);
+      }
+      let framerate_den = framerate_den as i32;
+      let framerate_num = framerate_num as i32;
+      let duration = (*stream).time_base.den * framerate_den
+        / ((*stream).time_base.num * framerate_num);
+      self.duration = duration as i64;
+    }
+  }
+
+  fn write_frame(&mut self, pts: u64, data: &[u8]) {
+    unsafe {
+      let mut pkt: AVPacket = mem::zeroed();
+      av_init_packet(&mut pkt);
+      pkt.data = data.as_ptr() as *mut _;
+      pkt.size = data.len() as i32;
+
+      let pts = pts as i64;
+      pkt.pts = pts * self.duration;
+      //no b frame, pts equals dts
+      pkt.dts = pts * self.duration;
```

Comment:
This abuses the fact that rav1e's timestamps will always increment by one. Is this OK?

Reply:
no, the name is not accurate, you can think duration is time unit. PTS is a number of the unit.
In our special case, one time unit is 1/fps.

Reply:
Which is exactly what I said above; it's a special case from the way rav1e handles timestamps. I just want to make sure it's OK to assume this will always be the case.

Reply:
Not clearly understand the question. Let explain more on the mechanism.
1. mp4 has a mp4_time_scale,  If it's 1000 it means 1s has 1000 mp4 time unit. 
2. rav1e application will give an fps to the muxer. 
3. We use fps as the video_time_scale. 
4. We set duration as mp4_time_scale/video_time_scale. suppose mp4_time_scale is 100, video_time_scale is 25. it means 1 video time unit has 40(1000/25) mp4 time unit.
5. Pts based on video timescale. we use pts*duration to convert it to mp4 timescale.
If we do not want to use fps as video_time_scale. We need to add new params to muxer and change step 3. But I do not think it's necessary, because use fps as the video time scale is perfectly well for video 
thanks

Reply:
In certain applications could produce a stream with timestamps like this:
```
0 10 25 26 28 29 32 42
```

Right now it does not happen because our input source cannot behave like that.

Once we accept input with arbitrary timestamps we'll have to deal with it.

Reply:
Yes, at some point we will be accepting proper variable framerate input and outputting it as such.